### PR TITLE
Use id from erc.yml

### PR DESCRIPTION
**Pull request by LukasLohoff**
The loader now uses the ID specified in `erc.yml` (#44 ) for compendia:

- Reads ID from erc.yml 
- Checks for validity and if a compendium with the ID already exists 
- Added tests for invalid ID & duplicate ID
- Reworked tests to reset database and local compendium storage before each test

### CODE REVIEW

File: lib/steps.js [hunk 1]
```diff
@@ -576,6 +577,108 @@ function detectCompendium(passon) {
...
+
+            // Get compendium yml, or throw exception on error
+            try {
+                let doc = yaml.safeLoad(fs.readFileSync(passon.detectionFile, 'utf8'));
+                compendiumID = doc.id;
+                debug('[%s] Successfully read compendium metadata: %O', passon.id, doc);
+
+            } catch (error) {
+                debug('[%s] Could not read compendium detection file %s: %s', passon.id, passon.detectionFile, error);
+                error.msg = 'Could not read compendium detection file';
```
Comment (nuest) [inline lib/steps.js]: You're not reading the compendium detection file, but the "compendium configuration file", see http://o2r.info/erc-spec/spec/#erc-configuration-file
 
Before it was only _used_ for detection. Please rephrase.

File: lib/steps.js [hunk 2]
```diff
@@ -576,6 +577,108 @@ function detectCompendium(passon) {
...
+            if (!compendiumID){
+                debug('[%s] No ID specified for compendium', passon.id);
+                let message = 'No id found in compendium detection file';
+                let error = new Error(message);
+                error.msg = message;
+                error.status = 400;
+                reject(error);
+            } else {
+                // Validate compendium id
+                let idRegex = new RegExp(config.bagtainer.id_regex);
```
Comment (nuest) [inline lib/steps.js]: I think `new RegExp` is not neccessary, you can directly do `config.bagtainer.id_regex.test(compendiumID)`.

File: lib/steps.js [hunk 3]
```diff
@@ -36,6 +36,8 @@ const util = require('util');
 const Docker = require('dockerode');
 const Stream = require('stream');
 const meta = require('../lib/meta');
+const yaml = require('js-yaml');
+const randomstring = require('randomstring');
```
Comment (nuest) [inline lib/steps.js]: `randomstring` is not used in this file AFAICS

File: lib/steps.js [hunk 4]
```diff
@@ -576,6 +577,108 @@ function detectCompendium(passon) {
...
+                    Compendium.findOne({ id: compendiumID }).select('id user').exec((err, compendium) => {
+                        // eslint-disable-next-line no-eq-null, eqeqeq
+                        if (err) {
+                            debug('[%s] Error querying compendium with id %s', passon.id, compendiumID);
+                            err.message = 'Error querying compendium';
+                            err.status = 400;
+                            reject(err);
+                        }
+                        if (compendium === null) {
+                            debug('[%s] Assigned new ID %s to compendium', passon.id, compendiumID);
```
Comment (nuest) [inline lib/steps.js]: Please clarify here: you are assigning the id loaded from the configuration file, not a "new" one.

File: lib/steps.js [hunk 5]
```diff
@@ -576,6 +577,108 @@ function detectCompendium(passon) {
...
+            let updatedPath = path.join(config.fs.compendium, passon.id);
+            debug('[%s] Copying compendium files from %s to %s due to ID specified in compendium file', passon.id, passon.compendium_path, updatedPath);
+
+            let cmd = 'mv ' + passon.compendium_path + ' ' + updatedPath;
+            exec(cmd, (error, stdout, stderr) => {
+                if (error || stderr) {
+                    debug(error, stderr, stdout);
+                    let errors = error.message.split(':');
+                    let message = errorMessageHelper(errors[errors.length - 1]);
+                    error.msg = 'moving compendium files to new location failed: ' + message;
```
Comment (nuest) [inline lib/steps.js]: Have you thought about what do do with the files on an error here? Should they be deleted?

File: lib/steps.js [hunk 6]
```diff
@@ -576,6 +577,108 @@ function detectCompendium(passon) {
...
+    return new Promise((fulfill, reject) => {
+
+        if (passon.isCompendium){
+            let updatedPath = path.join(config.fs.compendium, passon.id);
+            debug('[%s] Copying compendium files from %s to %s due to ID specified in compendium file', passon.id, passon.compendium_path, updatedPath);
+
+            let cmd = 'mv ' + passon.compendium_path + ' ' + updatedPath;
+            exec(cmd, (error, stdout, stderr) => {
+                if (error || stderr) {
+                    debug(error, stderr, stdout);
```
Comment (nuest) [inline lib/steps.js]: Please add a short message to this debug statement. Printing all outputs is good, a short text helps identifying where the error happens.

File: package.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "o2r-loader",
-  "version": "0.10.0",
+  "version": "0.10.1",
```
Comment (nuest) [inline package.json]: This is more than just a bugfix, please set version to `0.11.0`

**Review by nuest — CHANGES_REQUESTED** [review]

**Comment by nuest** [comment]
The tests on Travis fail (https://travis-ci.org/o2r-project/o2r-loader/builds/358827922):

```
/loader/node_modules/bindings/bindings.js:88
        throw e
        ^
Error: Error relocating /loader/node_modules/detect-character-encoding/build/Release/icuWrapper.node: utrace_cleanup_59: symbol not found
    at Object.Module._extensions..node (module.js:681:18)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at bindings (/loader/node_modules/bindings/bindings.js:81:44)
    at Object.<anonymous> (/loader/node_modules/detect-character-encoding/index.js:3:37)
    at Module._compile (module.js:652:30)
    at Object.Module._extensions..js (module.js:663:10)
    at Module.load (module.js:565:32)
    at tryModuleLoad (module.js:505:12)
    at Function.Module._load (module.js:497:3)
    at Module.require (module.js:596:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/loader/lib/steps.js:34:33)
```

@LukasLohoff You did not change the Dockerfile, so this surprises me... it seems to be the same issue we had with https://github.com/sonicdoe/detect-character-encoding/issues/8

- [ ] try out locally with with the current base image `node:8-alpine`
- [ ] try out locally with base image `node:9-alpine` (should not make a different, both use `alpine:3.6`!)
- [x] try out locall with `node:8-slim`

File: lib/steps.js (hunk 5 above)
Comment (LukasLohoff) [reply lib/steps.js]: I don't think it's necessary to delete the directory as it's just a directory with a randomly generated name (from the ID)

**Review by nuest — APPROVED** [review]

**Comment by nuest** [comment]
`node:8-slim` seems to have worked, but now we have troubles with the test setup, the test code cannot delete the files created by the microservice containers because the travis user is not root...

File: Dockerfile [hunk 8]
```diff
@@ -12,41 +12,27 @@
...
-  && rm -r /root/.cache
-
-# Add Alpine mirrors, replacing default repositories with edge ones, based on https://github.com/jfloff/alpine-python/blob/master/3.4/Dockerfile
-RUN echo "http://dl-cdn.alpinelinux.org/alpine/edge/testing" > /etc/apk/repositories \
-  && echo "http://dl-cdn.alpinelinux.org/alpine/edge/community" >> /etc/apk/repositories \
-  && echo "http://dl-cdn.alpinelinux.org/alpine/edge/main" >> /etc/apk/repositories
-
-RUN apk add --no-cache \
+RUN apt-get update && apt-get install -y \
+    python \
```
Comment (nuest) [inline Dockerfile:20]: Does this install Python 2 or 3? IIRC this is used for npm package `bagit`, just in case one of the tests with bags fails now.

Comment (LukasLohoff) [reply Dockerfile:20]: It installs python 2. `bagit` should work, I verified that in a separate `node:8-slim` container

**Pull request merged**